[WFLY-4916] Adds the Red Hat Support Lib as well as the Insights Subsystem

Kinlaw opened the pull request:
This pull request is for the addition of the Red Hat Support Lib and the Insights subsystem.

The Red Hat Support Lib is a Java API to the Red Hat subscriber REST services on redhat.com. Developers interested in leveraging redhat.com REST services can easily integrate them into a web application with this API.

The Insights subsystem is a service that allows the user to send the JDR to insights periodically. Red Hat will analyze the information and provide results related to security vulnerabilities, misconfigurations, performance issues, and other problems discovered in the received metadata. The service is disabled by default and must be manually enabled by the user.

Enabling Insights subsystem from the CLI:

1) Add subsystem to extensions in standalone.xml
    &lt;extension  module="org.jboss.as.insights"/&gt;
2) Add subsystem to submodules in standalone.xml
    &lt;subsystem xmlns="org.jboss.as.insights:1.0"/&gt;
3) Start standalone server and connect to CLI
4) To enable execute the following command from the CLI
    ./subsystem=insights:enable(rhnUid=[RHN-UID],rhnPw=[RHN-PASSWORD])
5) To change the frequency that the report is sent off
    ./subsystem=insights:write-attribute(name=frequency,value=[NUMBER-OF-DAYS]) where NUMBER-OF-DAYS is the amount of time that passes between each occurrence.

JIRA: https://issues.jboss.org/browse/WFLY-4916

Commit:
adds telemetry subsystem which currently consists of a timer that takes in the frequency at which the thread should run and performs that task each time the period completes itself

Commit:
adds telemetry subsystem which currently consists of a timer that takes in the frequency at which the thread should run and performs that task each time the period completes itself

Commit:
# This is a combination of 3 commits.
# The first commit's message is:
adds telemetry subsystem which currently consists of a timer that takes in the frequency at which the thread should run and performs that task each time the period completes itself

# This is the 2nd commit message:

updating poms

# This is the 3rd commit message:

removes added jars from feature-pack and updates redhat-support-lib pom to update jaxb2 version

Commit:
adds telemetry subsystem which currently consists of a timer that takes in the frequency at which the thread should run and performs that task each time the period completes itself

Commit:
# This is a combination of 3 commits.
# The first commit's message is:
adds telemetry subsystem which currently consists of a timer that takes in the frequency at which the thread should run and performs that task each time the period completes itself

# This is the 2nd commit message:

updating poms

# This is the 3rd commit message:

removes added jars from feature-pack and updates redhat-support-lib pom to update jaxb2 version

Commit:
generates jdr report inside of timer; frequency is measured in days

Commit:
generates jdr report inside of timer; frequency is measured in days

Commit:
generates jdr report inside of timer; frequency is measured in days

Commit:
adds jdr dependency to module.xml

Commit:
adds jdr dependency to module.xml

Commit:
adds jdr dependency to module.xml

Commit:
adds redhat-support-lib

Commit:
removing redhat-support-lib

Commit:
removing redhat-support-lib

Commit:
adding redhat-support-lib

Commit:
adding redhat-support-lib

Commit:
adding redhat-support-lib

Commit:
adding redhat-support-lib

Commit:
fixes build and commons-net module.xml

Commit:
updating poms

Commit:
updating poms

Commit:
updating poms

Commit:
removes added jars from feature-pack and updates redhat-support-lib pom to update jaxb2 version

Commit:
cleaning up BaseQuery

Commit:
cleaning up BaseQuery

Commit:
cleaning up BaseQuery

Commit:
cleaning up BaseQuery

Commit:
cleaning up file for checkstyle plugin

Commit:
cleaning up file for checkstyle plugin

Commit:
cleaning up file for checkstyle plugin

Commit:
cleaning up file for checkstyle plugin

Commit:
fixes xjc plugin

Commit:
fixes xjc plugin

Commit:
fixes xjc plugin

Commit:
fixes xjc plugin

Commit:
adds enable(rhnUid,rhnPw) jboss cli function for enabling telemetry with RHN user ID and PW

Commit:
updates module.xml to current state

Commit:
updates module.xml to current state

Commit:
updates module.xml to current state

Commit:
updates module.xml to current state

Commit:
changes classloader to properly load RESTEasy writers and readers

Commit:
adds services='import' to module.xml and manually adds writers to the RESTEasy client

Commit:
adds services='import' to module.xml and manually adds writers to the RESTEasy client

Commit:
adds services='import' to module.xml and manually adds writers to the RESTEasy client

Commit:
adds services='import' to module.xml and manually adds writers to the RESTEasy client

Commit:
changes classloader to properly load RESTEasy writers and readers

Commit:
switches classloader to RESTEasy classloader in order to add writers to RESTEasy client

Commit:
switches classloader to RESTEasy classloader in order to add writers to RESTEasy client

Commit:
switches classloader to RESTEasy classloader in order to add writers to RESTEasy client

Commit:
switches classloader to RESTEasy classloader in order to add writers to RESTEasy client

Commit:
moves ClassLoader switch down to where client is built

Commit:
moves ClassLoader switch down to where client is built

Commit:
moves ClassLoader switch down to where client is built

Commit:
moves ClassLoader switch down to where client is built

Commit:
Fixes runtime issues with classloaders and module.xml

Commit:
Fixes runtime issues with classloaders and module.xml

Commit:
Fixes runtime issues with classloaders and module.xml

Commit:
changes name of setWriters to setProviders since it is setting more than just writers

Commit:
changes name of setWriters to setProviders since it is setting more than just writers

Commit:
changes name of setWriters to setProviders since it is setting more than just writers

Commit:
changes name of setWriters to setProviders since it is setting more than just writers

Commit:
changing classloader switch to this instead of a specific library

Commit:
changing classloader switch to this instead of a specific library

Commit:
changing classloader switch to this instead of a specific library

Commit:
changing classloader switch to this instead of a specific library

Commit:
removing unused files

Commit:
removing unused files

Commit:
removing unused files

Commit:
removing unused files

Commit:
adds insights system check and creates new system if current systems UUID is not in insights; uploads jdr to insights for system

Commit:
updates error handling to use recommended JBoss logger

Commit:
adds TelemetryLogger and log statements to go with it for Telemetry

Commit:
adds TelemetryLogger and log statements to go with it for Telemetry

Commit:
updates error handling to use recommended JBoss logger

Commit:
updates error handling to use recommended JBoss logger

Commit:
adds TelemetryLogger and log statements to go with it for Telemetry

Commit:
updates error handling to use recommended JBoss logger

Commit:
removes properties from standalone.xml; implements the ability to create a new system if it is un_registered or it does not exist

Commit:
removes properties from standalone.xml; implements the ability to create a new system if it is un_registered or it does not exist

Commit:
removes properties from standalone.xml; implements the ability to create a new system if it is un_registered or it does not exist

Commit:
fixes build and commons-net module.xml

Commit:
fixes build and commons-net module.xml

Commit:
removes added jars from feature-pack and updates redhat-support-lib pom to update jaxb2 version

Commit:
removes added jars from feature-pack and updates redhat-support-lib pom to update jaxb2 version

Commit:
# This is a combination of 2 commits.
# The first commit's message is:

adds enable(rhnUid,rhnPw) jboss cli function for enabling telemetry with RHN user ID and PW

# This is the 2nd commit message:

updates module.xml to current state

Commit:
adds enable(rhnUid,rhnPw) jboss cli function for enabling telemetry with RHN user ID and PW

Commit:
adds enable(rhnUid,rhnPw) jboss cli function for enabling telemetry with RHN user ID and PW

Commit:
adds insights system check and creates new system if current systems UUID is not in insights; uploads jdr to insights for system

Commit:
adds insights system check and creates new system if current systems UUID is not in insights; uploads jdr to insights for system

Commit:
adds insights system check and creates new system if current systems UUID is not in insights; uploads jdr to insights for system

Commit:
adds insights system check and creates new system if current systems UUID is not in insights; uploads jdr to insights for system

Commit:
updates error handling to use recommended JBoss logger

Commit:
updates error handling to use recommended JBoss logger

Commit:
adds dependencies back in which were removed from master fetch

Commit:
adds dependencies back in which were removed from master fetch

Commit:
fixing merge conflicts

Commit:
fixing merge conflicts

Commit:
fixes merge conflict

Commit:
fixing merge conflicts

Commit:
fixing merge conflict

Commit:
fixing merge conflict

Commit:
fixing merge conflict

Commit:
fixing merge conflict

Commit:
fixing merge conflict

Commit:
fixing merge conflict

Commit:
fixing merge conflict

Commit:
fixing merge conflict

Commit:
merging

Commit:
fixing merge results

Commit:
adding in UUID from JDR

Commit:
adding in UUID from JDR

Commit:
adding in UUID from JDR

Commit:
removes direct reference to Telemetries; adds proxy variables to enable call

Commit:
removes direct reference to Telemetries; adds proxy variables to enable call

Commit:
adds logger for redhat-support-lib

Commit:
moves thread logic into its own try/catch block

Commit:
moves thread logic into its own try/catch block

Commit:
rename telemetry to insights

Commit:
rename telemetry to insights

Commit:
rename telemetry to insights

Commit:
rename telemetry to insights

Commit:
replaces telemetry with insights

Commit:
removes testing value for MILLISECOND_TO_DAY

Commit:
removes println statements

kabir committed:
Merge pull request #7756 from pferraro/master

Replace DefaultValueAttributeConverter with impl in wildfly-core

ctomc commented:
can you please squash this commits into something more reasonable? maybe have 5 commits that are logical units of work not 10+ commits that are "fixing merge conflicts"

see http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html for help on how to do it.

ctomc commented:
this is ok to test

ctomc commented on insights/pom.xml:
```diff
@@ -0,0 +1,148 @@
+                                <copy file="target/${project.artifactId}.jar" tofile="${dist_dir}/jboss-as-insights.jar" />
+
+                                <echo>Module ${module.name} has been created in the target/module directory. Copy to your WildFly installation.</echo>
+                            </target>
```
all this cannot be part of the build if you are inside wildfly codebase

ctomc commented on insights/pom.xml:
```diff
@@ -0,0 +1,148 @@
+        <dependency>
+            <groupId>com.redhat.gss</groupId>
+            <artifactId>wildfly-redhat-support-lib</artifactId>
+            <version>${project.version}</version>
```
that is bad practice, versions should be managed in parent

ctomc commented on insights/src/main/java/org/jboss/as/telemetry/extension/InsightsRequestHandler.java:
```diff
@@ -0,0 +1,76 @@
+        final String rhnPw = operation.require(InsightsExtension.RHNPW).asString();
+
+        if(operation.has(InsightsExtension.PROXY_PASSWORD) && operation.has(InsightsExtension.PROXY_USER)) {
+            String proxyPwd = operation.get(InsightsExtension.PROXY_PASSWORD).asString();
```
this should be InsightsSubsystemDefinition.RHNPW.resolveModelAttribute(operation, context)
and similarly all others.

ctomc commented on insights/src/main/java/org/jboss/as/telemetry/extension/InsightsService.java:
```diff
@@ -0,0 +1,490 @@
+        File file = new File(propertiesFilePath);
+        if (file.exists()) {
+            try {
+                fis = new FileInputStream(propertiesFilePath);
```
it would be easier if you would use try with resources here.
try (FileInputStream fis  = new FileInputStream(propertiesFilePath){
this way you don't need to manually close the stream

ctomc commented on insights/src/main/java/org/jboss/as/telemetry/extension/InsightsService.java:
```diff
@@ -0,0 +1,490 @@
+        int proxyPort = 0;
+        String userAgent = null;
+        String jbossConfig = System.getProperty(JBOSS_PROPERTY_DIR);
+        String propertiesFilePath = jbossConfig + File.separator
```
what is this config file?
and why is this not part of subsystem config.
any external configuration file makes it hard to mange / use subsystems in domain environment.

ctomc commented on insights/src/main/java/org/jboss/as/telemetry/extension/InsightsService.java:
```diff
@@ -0,0 +1,490 @@
+            ROOT_LOGGER.couldNotUploadJdr(e);
+        } catch (RequestException e) {
+            wasSuccessful = false;
+            ROOT_LOGGER.couldNotUploadJdr(e);
```
you could use catch mutliple given that you do same thing for different exceptions.
 } catch (RequestException|ParseException|MalformedURLException e) {

ctomc commented on insights/src/main/java/org/jboss/as/telemetry/extension/InsightsSubsystemDefinition.java:
```diff
@@ -0,0 +1,103 @@
+            new SimpleAttributeDefinitionBuilder(InsightsExtension.PROXY_USER, ModelType.STRING)
+                    .setAllowExpression(true)
+                    .setFlags(AttributeAccess.Flag.RESTART_ALL_SERVICES)
+                    .setDefaultValue(new ModelNode(""))
```
you really want it an empty string not undefined?

ctomc commented on insights/src/main/java/org/jboss/as/telemetry/extension/SubsystemAdd.java:
```diff
@@ -0,0 +1,49 @@
+    @Override
+    protected void populateModel(ModelNode operation, ModelNode model) throws OperationFailedException {
+        InsightsSubsystemDefinition.FREQUENCY.validateAndSet(operation, model);
+        InsightsSubsystemDefinition.ENABLED.validateAndSet(operation, model);
```
there are more attributes registered on resource, why you populate only two?

ctomc commented on build/assembly-src.xml:
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<assembly xmlns="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
this whole assembly is irrelevant for wildfly build

ctomc commented on feature-pack/src/main/resources/modules/system/layers/base/javax/net/ssl/main/module.xml:
```diff
@@ -0,0 +1,29 @@
+
+<module xmlns="urn:jboss:module:1.3" name="javax.net.ssl">
+    <dependencies>
+        <module name="javax.net.ssl" export="true"/>
```
What is this? module named javax.net.ssl that depends on javax.net.ssl?

ctomc commented on insights/src/main/java/org/jboss/as/telemetry/extension/InsightsExtension.java:
```diff
@@ -0,0 +1,116 @@
+        public void readElement(XMLExtendedStreamReader reader, List<ModelNode> list) throws XMLStreamException {
+            // Require no attributes
+            ParseUtils.requireNoAttributes(reader);
+            ParseUtils.requireNoContent(reader);
```
subsystem parser that doesn't parse any attributes that are defined in the subsystem?
that is wrong.

ctomc commented on insights/src/main/resources/schema/insights.xsd:
```diff
@@ -0,0 +1,9 @@
+            attributeFormDefault="unqualified"
+            version="1.0">
+
+    <xs:element name="insights" type="subsystemType"/>
```
parser doesn't handle this at all.

ctomc commented:
build failure:
```
[11:51:13][Step 2/3] [ERROR] The build could not read 2 projects -> [Help 1]
[11:51:13][Step 2/3] [ERROR]   
[11:51:13][Step 2/3] [ERROR]   The project org.wildfly:wildfly-insights:10.0.0.Alpha5-SNAPSHOT (/opt/buildAgent/work/a31d203e70e89f90/insights/pom.xml) has 1 error
[11:51:13][Step 2/3] [ERROR]     'dependencies.dependency.version' for commons-net:commons-net:jar is missing. @ org.wildfly:wildfly-insights:[unknown-version], /opt/buildAgent/work/a31d203e70e89f90/insights/pom.xml, line 78, column 21
[11:51:13][Step 2/3] [ERROR]   
[11:51:13][Step 2/3] [ERROR]   The project com.redhat.gss:wildfly-redhat-support-lib:10.0.0.Alpha5-SNAPSHOT (/opt/buildAgent/work/a31d203e70e89f90/redhat-support-lib/pom.xml) has 1 error
[11:51:13][Step 2/3] [ERROR]     'dependencies.dependency.version' for commons-net:commons-net:jar must be a valid version but is '${version.commons-net}'. @ com.redhat.gss:wildfly-redhat-support-lib:[unknown-version], /opt/buildAgent/work/a31d203e70e89f90/redhat-support-lib/pom.xml, line 66, column 26
```

ctomc commented:
Also on top of all this, I think this shouldn't be part of WildFly codebase.
It would be much better if it would be a feature pack that WildFly/EAP distribution would just pull in.
So it can be maintained in separate repository and have its own release cycle.

ctomc commented:
subsystem is missing test cases.

Commit:
fixes xjc plugin

Commit:
# This is a combination of 6 commits.
# The first commit's message is:
mergining

# This is the 2nd commit message:

switches classloader to RESTEasy classloader in order to add writers to RESTEasy client

# This is the 3rd commit message:

moves ClassLoader switch down to where client is built

# This is the 4th commit message:

Fixes runtime issues with classloaders and module.xml

# This is the 5th commit message:

changes name of setWriters to setProviders since it is setting more than just writers

# This is the 6th commit message:

changing classloader switch to this instead of a specific library

Commit:
merging

Kinlaw commented:
@ctomc Thanks for taking a detailed look at the changes. I am working on making the corrections to the issues pointed out in your comments. I agree that it would be better if the Red Hat Support Lib lived outside of WildFly. The codebase already exists on GitHub [1] but the changes specific to WildFly will need to be added to it. I'm not sure how to import it into WildFly. How should I go about doing this?

[1] https://github.com/redhataccess/redhat-support-lib-java

bstansberry commented on insights/pom.xml:
```diff
@@ -0,0 +1,148 @@
+                  projects that depend on this project.-->
+            <scope>provided</scope>
+            <optional>true</optional>
+        </dependency>
```
These last two are duplicates.

bstansberry commented on insights/src/main/java/org/jboss/as/telemetry/extension/InsightsEnabledHandler.java:
```diff
@@ -0,0 +1,45 @@
+package org.jboss.as.insights.extension;
```
Copyright header.

bstansberry replied:
Please assume that comment is on every class file in the PR where it doesn't exist.

bstansberry commented on insights/src/main/java/org/jboss/as/telemetry/extension/InsightsEnabledHandler.java:
```diff
@@ -0,0 +1,45 @@
+     */
+    protected void revertUpdateToRuntime(OperationContext context, ModelNode operation, String attributeName,
+          ModelNode valueToRestore, ModelNode valueToRevert, Void handback) {
+        // no-op
```
Why a no-op?

bstansberry commented on insights/src/main/resources/schema/insights.xsd:
```diff
@@ -0,0 +1,9 @@
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema"
```
We have a standard naming scheme for the subsystem config xsd files we ship. This file should be named:

wildfly-insights_1_0.xsd

Since this won't actually end up in the WildFly dist, this is fine:

jboss-eap-insights_1_0.xsd

bstansberry commented on insights/src/main/java/org/jboss/as/telemetry/extension/InsightsExtension.java:
```diff
@@ -0,0 +1,116 @@
+    protected static final String FREQUENCY = "frequency";
+    protected static final String RHNPW = "rhnPw";
+    protected static final String RHNUID = "rhnUid";
+    public static final String PROXY_USER = "proxyUser";
```
EAP schema elements/attributes and management resource addresses/attributes do not use camel case. A conforming name would be

proxy-user

Same comment applies to the other attributes.

bstansberry commented on insights/src/main/java/org/jboss/as/telemetry/extension/InsightsFrequencyHandler.java:
```diff
@@ -0,0 +1,46 @@
+     */
+    protected void revertUpdateToRuntime(OperationContext context, ModelNode operation, String attributeName,
+          ModelNode valueToRestore, ModelNode valueToRevert, Void handback) {
+        // no-op
```
Why no-op?

bstansberry commented on insights/src/main/java/org/jboss/as/telemetry/extension/InsightsService.java:
```diff
@@ -0,0 +1,490 @@
+    public static final String USERNAME = "username";
+    public static final String PASSWORD = "password";
+    public static final String URL = "url";
+    public static final String USER_AGENT = "userAgent";
```
No camel case.

bstansberry commented on insights/src/main/java/org/jboss/as/telemetry/extension/InsightsService.java:
```diff
@@ -0,0 +1,490 @@
+                fis = new FileInputStream(file.getPath());
+                properties.load(fis);
+            }
+            properties.setProperty(InsightsExtension.PROXY_PASSWORD, proxyPwd);
```
Security hole. Plain text credential written to disk.

bstansberry commented on insights/src/main/java/org/jboss/as/telemetry/extension/InsightsService.java:
```diff
@@ -0,0 +1,490 @@
+                properties.load(fis);
+            }
+            properties.setProperty(InsightsExtension.PROXY_PASSWORD, proxyPwd);
+            properties.setProperty(InsightsExtension.PROXY_USER, proxyUser);
```
Security hole. Plain text credential written to disk.

bstansberry commented:
The fundamental question here is why is the subsystem configuration not being persisted to the configuration document, but instead is written to a separate file? What led to this choice?

bstansberry commented:
Once the various questions are sorted, this is going to need to be squashed down to a reasonable number of commits. By chance Jason Greene had a pretty good comment on a different PR re: multiple commits, so I'll quote it here:

```
In general I think its ok if there are multiple commits as long as

1) They can pass the test suite in isolation, such that git bisect works
2) They are discrete well contained units of work
3) There is a very small number of them
4) They are related
```

We're not exactly rigid about that, but that's the general spirit and our commit history very much reflects that.

Commit:
merging

Commit:
Merge branch 'WFLY-4916' of github.com:Kinlaw/wildfly into WFLY-4916

Commit:
updates wildfly version to ALPHA6

Commit:
Merge branch 'WFLY-4916' of github.com:Kinlaw/wildfly into WFLY-4916

Commit:
Implements suggested changes indicated by comments in PR7752

Commit:
renames classpath to match package

Commit:
moves redhat-support-lib-java outside of WildFly project

Commit:
moves redhat-support-lib-java outside of WildFly project

Commit:
Merge branch 'WFLY-4916' of github.com:Kinlaw/wildfly into WFLY-4916

Commit:
updates redhat-support-lib version

jamezp commented:
The issue is you're parent is still set to `10.0.0.Alpha6-SNAPSHOT` and master is on `10.0.0.Beta2-SNAPSHOT`. Please update the parent so it will build.

Kinlaw commented:
Thanks, James. Will do.

----- Original Message -----
> From: "James R. Perkins" <notifications@github.com>
> To: "wildfly/wildfly" <wildfly@noreply.github.com>
> Cc: "Josh Kinlaw" <jkinlaw@redhat.com>
> Sent: Thursday, August 20, 2015 4:09:03 PM
> Subject: Re: [wildfly] [WFLY-4916] Adds the Red Hat Support Lib as well as the Insights Subsystem (#7752)
> 
> The issue is you're parent is still set to `10.0.0.Alpha6-SNAPSHOT` and
> master is on `10.0.0.Beta2-SNAPSHOT`. Please update the parent so it will
> build.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/wildfly/wildfly/pull/7752#issuecomment-133157059

bstansberry commented:
Hi Josh,

This PR is messing with our CI job, somehow triggering continual tries to test it. Please open a new PR once you have a version that's squashed and builds locally and that Emmanuel can build locally as well.
